plugins/ndpi/ndpi.c

```diff
@@ -0,0 +1,542 @@
+};
+
+struct NdpiFlowContext {
+    struct ndpi_flow_struct *ndpi_flow;
+    ndpi_protocol detected_l7_protocol;
+    uint8_t detection_completed;
+};
+
+typedef struct DetectnDPIProtocolData_ {
+    ndpi_master_app_protocol l7_protocol;
+    uint8_t negated;
+} DetectnDPIProtocolData;
+
+typedef struct DetectnDPIRiskData_ {
+    ndpi_risk risk_mask; /* uint64 */
+    uint8_t negated;
+} DetectnDPIRiskData;
+
+static void ThreadStorageFree(void *ptr)
+{
+    SCLogDebug("Free'ing nDPI thread storage");
+    struct NdpiThreadContext *context = ptr;
+    ndpi_exit_detection_module(context->ndpi);
+    SCFree(context);
+}
+
+static void FlowStorageFree(void *ptr)
+{
+    struct NdpiFlowContext *ctx = ptr;
+    ndpi_flow_free(ctx->ndpi_flow);
+    SCFree(ctx);
+}
+
+static void OnFlowInit(ThreadVars *tv, Flow *f, const Packet *p, void *_data)
+{
+    struct NdpiFlowContext *flowctx = SCCalloc(1, sizeof(*flowctx));
+    if (flowctx == NULL) {
+        FatalError("Failed to allocate nDPI flow context");
+    }
+
+    flowctx->ndpi_flow = ndpi_flow_malloc(SIZEOF_FLOW_STRUCT);
+    if (flowctx->ndpi_flow == NULL) {
+        FatalError("Failed to allocate nDPI flow");
+    }
+
+    memset(flowctx->ndpi_flow, 0, SIZEOF_FLOW_STRUCT);
+    flowctx->detection_completed = 0;
+    FlowSetStorageById(f, flow_storage_id, flowctx);
+}
+
+static void OnFlowUpdate(ThreadVars *tv, Flow *f, Packet *p, void *_data)
+{
+    struct NdpiThreadContext *threadctx = ThreadGetStorageById(tv, thread_storage_id);
+    struct NdpiFlowContext *flowctx = FlowGetStorageById(f, flow_storage_id);
+    uint16_t ip_len = 0;
+    void *ip_ptr = NULL;
+
+    if (!threadctx->ndpi || !flowctx->ndpi_flow) {
+        return;
+    }
+
+    if (PacketIsIPv4(p)) {
+        const IPV4Hdr *ip4h = PacketGetIPv4(p);
+        ip_len = IPV4_GET_RAW_IPLEN(ip4h);
+        ip_ptr = (void *)PacketGetIPv4(p);
+    } else if (PacketIsIPv6(p)) {
+        const IPV6Hdr *ip6h = PacketGetIPv6(p);
+        ip_len = IPV6_HEADER_LEN + IPV6_GET_RAW_PLEN(ip6h);
+        ip_ptr = (void *)PacketGetIPv6(p);
+    }
+
+    if (!flowctx->detection_completed && ip_ptr != NULL && ip_len > 0) {
+        uint64_t time_ms = ((uint64_t)p->ts.secs) * 1000 + p->ts.usecs / 1000;
+
+        SCLogDebug("Performing nDPI detection...");
+
+        flowctx->detected_l7_protocol = ndpi_detection_process_packet(
+                threadctx->ndpi, flowctx->ndpi_flow, ip_ptr, ip_len, time_ms, NULL);
+
+        if (ndpi_is_protocol_detected(flowctx->detected_l7_protocol) != 0) {
+            if (!ndpi_is_proto_unknown(flowctx->detected_l7_protocol.proto)) {
+                if (!ndpi_extra_dissection_possible(threadctx->ndpi, flowctx->ndpi_flow))
+                    flowctx->detection_completed = 1;
+            }
+        } else {
+            u_int16_t max_num_pkts = (f->proto == IPPROTO_UDP) ? 8 : 24;
+
+            if ((f->todstpktcnt + f->tosrcpktcnt) > max_num_pkts) {
+                u_int8_t proto_guessed;
```

Comment:
uint8_t

Needs a full scan for the `u_` variant of int types. We use the one w/o underscore.